frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+		}
+
+		/// Migrate the current top key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_top(&mut self) {
+			let last_top = self.last_top.as_ref().expect("value checked to be `Some`; qed");
+			let added_size = if let Some(data) = sp_io::storage::get(&last_top) {
+				self.dyn_size = self.dyn_size.saturating_add(data.len() as u32);
+				sp_io::storage::set(last_top, &data);
+				data.len() as u32
+			} else {
+				Zero::zero()
+			};
+
+			self.dyn_top_items.saturating_inc();
+			let next_key = sp_io::storage::next_key(last_top);
+			self.last_top = next_key;
+			log!(trace, "migrated a top key with size {}, next_task = {:?}", added_size, self);
+		}
+	}
+
+	/// The limits of a migration.
+	#[derive(Clone, Copy, Encode, Decode, scale_info::TypeInfo, Default, Debug, PartialEq, Eq)]
+	pub struct MigrationLimits {
+		/// The byte size limit.
+		pub size: u32,
+		/// The number of keys limit.
+		pub item: u32,
+	}
+
+	/// How a migration was computed.
+	#[derive(Clone, Copy, Encode, Decode, scale_info::TypeInfo, Debug, PartialEq, Eq)]
+	pub enum MigrationCompute {
+		/// A signed origin triggered the migration.
+		Signed,
+		/// An unsigned origin triggered the migration.
+		Unsigned,
+		/// An automatic task triggered the migration.
+		Auto,
+	}
+
+	/// Inner events of this pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Given number of `(top, child)` keys were migrated respectively, with the given
+		/// `compute`.
+		Migrated(u32, u32, MigrationCompute),
+	}
+
+	/// The outer Pallet struct.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Configurations of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>> {
+		/// Origin that can control the configurations of this pallet.
+		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::Origin>;
+
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The currency provider type.
+		type Currency: Currency<Self::AccountId>;
+
+		/// The amount of deposit collected per item in advance, for signed migrations.
+		///
+		/// This should reflect the average storage value size in the worse case.
+		type SignedDepositPerItem: Get<BalanceOf<Self>>;
+
+		/// The maximum limits that the signed migration could use.
+		type SignedMigrationMaxLimits: Get<MigrationLimits>;
+
+		/// The weight information of this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The priority used for unsigned transactions.
+		type UnsignedPriority: Get<TransactionPriority>;
+
+		/// The number of items that offchain worker will subtract from the first item count that
+		/// causes an over-consumption.
+		///
+		/// A value around 5-10 is reasonable.
+		type UnsignedBackOff: Get<u32>;
```

Comment:
I think it is nowhere explained how to get the correct number here. why 1 isn't always fine ?